eventsourcing/src/main/java/org/axonframework/eventsourcing/eventstore/jpa/LegacyJpaEventStorageEngine.java

```diff
@@ -0,0 +1,625 @@
+                    lastItem -> transactionManager.fetchInTransaction(
+                            () -> legacyJpaOperations.fetchDomainEvents(
+                                    identifier,
+                                    lastItem == null ? firstSequenceNumber : lastItem.getSequenceNumber() + 1,
+                                    lastSequenceNumber,
+                                    batchSize)
+                    )
+                    , finalAggregateBatchPredicate);
+            return StreamSupport.stream(spliterator, false);
+        }
+
+        Stream<? extends TrackedEventData<?>> readEventData(TrackingToken trackingToken) {
+            EventStreamSpliterator<? extends TrackedEventData<?>> spliterator = new EventStreamSpliterator<>(
+                    lastItem -> fetchTrackedEvents(lastItem == null ? trackingToken : lastItem.trackingToken(),
+                                                   batchSize),
+                    batch -> BATCH_OPTIMIZATION_DISABLED
+            );
+            return StreamSupport.stream(spliterator, false);
+        }
+
+        private List<? extends TrackedEventData<?>> fetchTrackedEvents(TrackingToken lastToken, int batchSize) {
+            Assert.isTrue(
+                    lastToken == null || lastToken instanceof GapAwareTrackingToken,
+                    () -> String.format("Token [%s] is of the wrong type. Expected [%s]",
+                                        lastToken, GapAwareTrackingToken.class.getSimpleName())
+            );
+
+            GapAwareTrackingToken previousToken = cleanedToken((GapAwareTrackingToken) lastToken);
+
+            List<Object[]> entries = transactionManager.fetchInTransaction(() -> legacyJpaOperations.fetchEvents(
+                    previousToken,
+                    batchSize));
+            return legacyJpaOperations.entriesToEvents(
+                    previousToken, entries, tokenOperations.gapTimeoutThreshold(), lowestGlobalSequence, maxGapOffset
+            );
+        }
+
+        private GapAwareTrackingToken cleanedToken(GapAwareTrackingToken lastToken) {
+            if (lastToken != null && lastToken.getGaps().size() > gapCleaningThreshold) {
+                return tokenOperations.withGapsCleaned(lastToken, indexAndTimestampBetweenGaps(lastToken));
+            }
+            return lastToken;
+        }
+
+        private List<Object[]> indexAndTimestampBetweenGaps(GapAwareTrackingToken lastToken) {
+            return transactionManager.fetchInTransaction(() -> legacyJpaOperations.indexAndTimestampBetweenGaps(
+                    lastToken));
+        }
+
+
+        private boolean defaultFinalAggregateBatchPredicate(List<? extends DomainEventData<?>> recentBatch) {
+            return fetchForAggregateUntilEmpty() ? recentBatch.isEmpty() : recentBatch.size() < batchSize;
+        }
+
+        private static class EventStreamSpliterator<T> extends Spliterators.AbstractSpliterator<T> {
```

Comment:
This is the spliterator from the `BatchingEventStorageEngine` I assume?

Reply:
Yes, just copy-paste. The old one will be removed with the containing class.

Reply:
Sounds good to me